fix(search): fixed search for carpooling according to doc

almerion opened the pull request:
closes [https://github.com/DUT-Info-Montreuil/SAE-5.A-carmate/issues/46](https://github.com/DUT-Info-Montreuil/SAE-5.A-carmate/issues/46)

naulan-chrzaszcz commented on src/database/repositories/carpooling_repository.py:
```diff
@@ -28,25 +29,25 @@ def get_carpoolings_route(self,
                         AND ABS(starting_point[2] - %s) < {self.RADIUS} 
                         AND ABS(destination[1] - %s) < {self.RADIUS}
                         AND ABS(destination[2] - %s) < {self.RADIUS}
-                        AND departure_date_time >= to_timestamp(%s)
```
Mé ? 😒

almerion replied:
Je peux le remettre mais ça plante ._.
pas besoin du to_timestamp

naulan-chrzaszcz commented on src/database/repositories/carpooling_repository.py:
```diff
@@ -28,25 +29,25 @@ def get_carpoolings_route(self,
                         AND ABS(starting_point[2] - %s) < {self.RADIUS} 
                         AND ABS(destination[1] - %s) < {self.RADIUS}
                         AND ABS(destination[2] - %s) < {self.RADIUS}
-                        AND departure_date_time >= to_timestamp(%s)
```
Pourquoi avoir enlever ca ?

almerion replied:
parceque ça plante idem que celle du dessus

naulan-chrzaszcz commented on src/database/repositories/carpooling_repository.py:
```diff
@@ -55,9 +56,9 @@ def get_carpoolings_route(self,
-                except ProgrammingError:
-                    raise NotFound("driver not found")
+                    curs.execute(query_nb_carpoolings_route, (start_lat, start_lon, end_lat, end_lon, str(departure_date_time),))
+                except ProgrammingError as e:
```
```suggestion
              except ProgrammingError:
```

naulan-chrzaszcz reviewed (changes requested):
Je comprend pas la raison d'enlever le `timestamp`, on va redevoir changer le code de mathéo... On peut toujours avoir la date sans prendre en compte l'heure...
Ca fait que tu mets des conditions en plus pour verifier si il y a une heures alors que datetime a probablement une fonction pour avoir uniquement l'heure du timestamp mit en argument d'URL

J'pense qu'il fallait juste ajuster la requête sql

naulan-chrzaszcz commented:
En soit c'est pas un fix mais un refactor car la route fonctionner de base

almerion commented:
Elle ne renvoyait pas les bonnes données et ne prennait pas les bonnes données en paramètre, un fix ducoup ..

naulan-chrzaszcz commented:
Ta rajouter un autre type de donnée a la route search c'est pas un fix

naulan-chrzaszcz commented on src/database/repositories/carpooling_repository.py:
```diff
@@ -28,36 +30,38 @@ def get_carpoolings_route(self,
-                    WHERE ABS(starting_point[1] - %s) < {self.RADIUS} 
+        query = f"""SELECT c.id, c.starting_point, c.destination, c.max_passengers, c.price, c.departure_date_time, c.driver_id
+                    FROM carmate.{self.POSTGRES_TABLE_NAME} as c 
+                    LEFT JOIN carmate.{ReserveCarpoolingRepository.POSTGRES_TABLE_NAME} as r on (c.id = r.carpooling_id)
```
```suggestion
                    LEFT JOIN carmate.{self.POSTGRES_TABLE_NAME} as r on (c.id = r.carpooling_id)
```

almerion replied:
Bruh non c'est pas la bonne table 🗿

naulan-chrzaszcz commented on src/database/repositories/carpooling_repository.py:
```diff
@@ -14,9 +16,9 @@ def get_carpoolings_route(self,
+                              departure_date_time: datetime,
                               page: int = 1,
-                              per_page: int = 10) -> Tuple[int, List[CarpoolingTable]]: ...
+                              per_page: int = 10) -> Tuple[int, List[CarpoolingForRecap]] | Tuple[int, List]: ...
```
Ca nique un peu les autres classe sur les valeurs retourné dans les fonctions de la classe repository ? C'est pas uniforme.
Qu'est-ce que t'en pense ?

naulan-chrzaszcz replied:
Tous les autres fonctions des autres repository renvoie XxxxxTable

almerion replied:
https://github.com/DUT-Info-Montreuil/SAE-5.A-carmate-back/blob/e19a9119574194b92587314ca96c3c38a8534cc4/src/database/repositories/license_repository.py#L24

Enfin après je peux toujours retourner le dto + une autre liste qui va associer carpooling_id & nb_places_reserves mais il y aura 3 select dans la fonction du coup

naulan-chrzaszcz replied:
> https://github.com/DUT-Info-Montreuil/SAE-5.A-carmate-back/blob/e19a9119574194b92587314ca96c3c38a8534cc4/src/database/repositories/license_repository.py#L24

Effectivement tu souligne un problème d'uniformité que les précèdents reviewer sur la PR n'étais pas capable de le trouvé 👹👹👹

> Enfin après je peux toujours retourner le dto + une autre liste qui va associer carpooling_id & nb_places_reserves mais il y aura 3 select dans la fonction du coup et le worker devra les mapper entre eux

Après c'est pas obliger, c'est qu'une suggestion

almerion replied:
très franchement moi ça me change rien de rien je le laisserait comme ça si c'était moi alors c'est comme tu le sens

almerion replied:
Si on voulait faire bien on devrait retourner aucune table mais que des dto de la db mais actuellement on renvoie que des tables

naulan-chrzaszcz replied:
Si c'est déjà fait sur la fonction `get_license_not_validated`, j'peux rien dire, laisse comme ca du coup

naulan-chrzaszcz commented on src/api/controller/carpooling/search.py:
```diff
@@ -27,18 +28,22 @@ def search_route_carpooling_api(self):
         end_lon: float
         try:
-            departure_date_time = int(request.args.get('departure_date_time'))
+            departure_date_time = datetime.strptime(request.args.get('departure_date_time'), "%Y-%m-%dT%H:%M:%S") \
+                if "T" in request.args.get('departure_date_time') \
+                else datetime.strptime(request.args.get('departure_date_time'), "%Y-%m-%d")
```
```suggestion
            departure_date_time = datetime.strptime(request.args.get('departure_date_time'), "%Y-%m-%dT%H:%M:%S" if "T" in request.args.get('departure_date_time') else "%Y-%m-%d")
```

naulan-chrzaszcz commented on src/api/controller/carpooling/search.py:
```diff
@@ -54,7 +59,8 @@ def search_route_carpooling_api(self):
                                                                                        departure_date_time,
                                                                                        page=page)
-        except Exception:
+        except Exception as e:
+            print(e)
```
```suggestion
        except Exception:
```

naulan-chrzaszcz commented on src/api/controller/carpooling/search.py:
```diff
@@ -1,6 +1,7 @@
 from flask import Blueprint, abort, jsonify, request
 
+from datetime import datetime
 from api.worker.carpooling.models import CarpoolingDTO
```
```suggestion

from api.worker.carpooling.models import CarpoolingDTO
```

naulan-chrzaszcz commented on src/api/controller/carpooling/search.py:
```diff
@@ -1,6 +1,7 @@
 from typing import List, Tuple
 from flask import Blueprint, abort, jsonify, request
 
+from datetime import datetime
```
```suggestion
from datetime import datetime

```

naulan-chrzaszcz commented on src/api/worker/carpooling/use_case/get_route_carpoolings.py:
```diff
@@ -18,16 +19,13 @@ def worker(self,
                page: int | None = None) -> Tuple[int, List[CarpoolingDTO]]:
         carpoolings_data: Tuple[int, List[CarpoolingTable]]
         try:
             if page is not None:
-                carpoolings_data = self.carpooling_repository.get_carpoolings_route(start_lat, start_lon, end_lat, end_lon, departure_date_time,
-                                                                                     page=page)
+                return self.carpooling_repository.get_carpoolings_route(start_lat, start_lon, end_lat, end_lon, departure_date_time,                                                                     page=page)
             else:
-                carpoolings_data = self.carpooling_repository.get_carpoolings_route(start_lat, start_lon, end_lat, end_lon, departure_date_time)
+                return self.carpooling_repository.get_carpoolings_route(start_lat, start_lon, end_lat, end_lon, departure_date_time)
         except Exception as e:
             raise InternalServerError(str(e))
-        return (carpoolings_data[0], 
-                [CarpoolingDTO.from_table(carpooling_table) for carpooling_table in carpoolings_data[1]])
```
Il y a un problème

naulan-chrzaszcz commented on src/api/worker/carpooling/use_case/get_route_carpoolings.py:
```diff
@@ -18,16 +19,13 @@ def worker(self,
-               departure_date_time: int,
+               departure_date_time: datetime,
                page: int | None = None) -> Tuple[int, List[CarpoolingDTO]]:
         carpoolings_data: Tuple[int, List[CarpoolingTable]]
```
```suggestion
```

naulan-chrzaszcz commented on src/api/worker/carpooling/use_case/get_route_carpoolings.py:
```diff
@@ -18,16 +19,13 @@ def worker(self,
                end_lon: float,
-               departure_date_time: int,
+               departure_date_time: datetime,
                page: int | None = None) -> Tuple[int, List[CarpoolingDTO]]:
```
```suggestion
               page: int | None = None) -> Tuple[int, List[CarpoolingTable]]:
```

naulan-chrzaszcz commented on src/database/repositories/carpooling_repository.py:
```diff
@@ -28,36 +30,38 @@ def get_carpoolings_route(self,
                               end_lat: float,
                               end_lon: float,
-                              departure_date_time: int,
+                              departure_date_time: datetime,
```
```suggestion
                              departure_date_time: int,
```

naulan-chrzaszcz reviewed (changes requested):
Des ajustements:
- La page 0 existe
- Code style
- Le `worker` de `get_route_carpoolings` n'est pas bien nettoyer

Des modifications injustifié:
![rock-one-eyebrow-raised-rock-staring](https://github.com/DUT-Info-Montreuil/SAE-5.A-carmate-back/assets/67024770/c70748db-d390-48d1-8296-0f846bbaac7a)
- J'ai besoin d'explication sur le fait d'avoir mit `datetime`, est-ce vraiment utile et pourquoi ? 
- Comment ta fait pour que ça plante ?
- Ta demandé a Mathéo si ça va pas niquer son code ?

naulan-chrzaszcz commented:
https://www.unixtimestamp.com/

almerion commented:
> Des ajustements:
> 
> * La page 0 existe
> * Code style
> * Le `worker` de `get_route_carpoolings` n'est pas bien nettoyer
> 
> Des modifications injustifié: ![rock-one-eyebrow-raised-rock-staring](https://private-user-images.githubusercontent.com/67024770/291075917-c70748db-d390-48d1-8296-0f846bbaac7a.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDI5MjU4NTIsIm5iZiI6MTcwMjkyNTU1MiwicGF0aCI6Ii82NzAyNDc3MC8yOTEwNzU5MTctYzcwNzQ4ZGItZDM5MC00OGQxLTgyOTYtMGY4NDZiYmFhYzdhLmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjE4VDE4NTIzMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWRjZDYxZmEzMTE5ZmYwYjJkZTc2ZTQ3NzM3NzA1MTA0MDBjYjRmYjY0NWQ5NjY1NDQwNTNlZmRhY2I3NzRmY2ImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.mzZIwse7roT7TipWeJdmK80oDWL8dW9B7B5Z0mq83Go) [ ](https://private-user-images.githubusercontent.com/67024770/291075917-c70748db-d390-48d1-8296-0f846bbaac7a.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDI5MjU4NTIsIm5iZiI6MTcwMjkyNTU1MiwicGF0aCI6Ii82NzAyNDc3MC8yOTEwNzU5MTctYzcwNzQ4ZGItZDM5MC00OGQxLTgyOTYtMGY4NDZiYmFhYzdhLmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjE4VDE4NTIzMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWRjZDYxZmEzMTE5ZmYwYjJkZTc2ZTQ3NzM3NzA1MTA0MDBjYjRmYjY0NWQ5NjY1NDQwNTNlZmRhY2I3NzRmY2ImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.mzZIwse7roT7TipWeJdmK80oDWL8dW9B7B5Z0mq83Go) ![rock-one-eyebrow-raised-rock-staring](https://private-user-images.githubusercontent.com/67024770/291075917-c70748db-d390-48d1-8296-0f846bbaac7a.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDI5MjU4NTIsIm5iZiI6MTcwMjkyNTU1MiwicGF0aCI6Ii82NzAyNDc3MC8yOTEwNzU5MTctYzcwNzQ4ZGItZDM5MC00OGQxLTgyOTYtMGY4NDZiYmFhYzdhLmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjE4VDE4NTIzMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWRjZDYxZmEzMTE5ZmYwYjJkZTc2ZTQ3NzM3NzA1MTA0MDBjYjRmYjY0NWQ5NjY1NDQwNTNlZmRhY2I3NzRmY2ImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.mzZIwse7roT7TipWeJdmK80oDWL8dW9B7B5Z0mq83Go) [ ](https://private-user-images.githubusercontent.com/67024770/291075917-c70748db-d390-48d1-8296-0f846bbaac7a.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDI5MjU4NTIsIm5iZiI6MTcwMjkyNTU1MiwicGF0aCI6Ii82NzAyNDc3MC8yOTEwNzU5MTctYzcwNzQ4ZGItZDM5MC00OGQxLTgyOTYtMGY4NDZiYmFhYzdhLmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjE4VDE4NTIzMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWRjZDYxZmEzMTE5ZmYwYjJkZTc2ZTQ3NzM3NzA1MTA0MDBjYjRmYjY0NWQ5NjY1NDQwNTNlZmRhY2I3NzRmY2ImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.mzZIwse7roT7TipWeJdmK80oDWL8dW9B7B5Z0mq83Go)
> 
> * J'ai besoin d'explication sur le fait d'avoir mit `datetime`, est-ce vraiment utile et pourquoi ?
> * Comment ta fait pour que ça plante ?
> * Ta demandé a Mathéo si ça va pas niquer son code ?

J'ai juste suivi la doc, si son code plante c'est que la doc n'a pas été suivie je suppose ._.
Mais ça plante pas ._.

naulan-chrzaszcz reviewed (changes requested):
Il y a deux trucs qui ont été mit en "resolve" mais rien n'a changé
 - https://github.com/DUT-Info-Montreuil/SAE-5.A-carmate-back/pull/37#discussion_r1429197607

almerion commented:
> Il y a deux trucs qui ont été mit en "resolve" mais rien n'a changé
> 
> * [fix(search): fixed search for carpooling according to doc #37 (comment)](https://github.com/DUT-Info-Montreuil/SAE-5.A-carmate-back/pull/37#discussion_r1429197607)
> * [fix(search): fixed search for carpooling according to doc #37 (comment)](https://github.com/DUT-Info-Montreuil/SAE-5.A-carmate-back/pull/37#discussion_r1429197800)

aucun des deux est en resolve

naulan-chrzaszcz commented on test/mock/carpooling/in_memory_carpooling_repository.py:
```diff
@@ -50,7 +52,10 @@ def get_carpoolings_route(
+        if self.reserve_carpooling_repository is None:
+            raise Exception("To use the fucntion get_carpoolings_route you need the reserve carpooling repository")
+
+        filtered_carpoolings = copy.deepcopy([
```
`deepcopy`?

naulan-chrzaszcz reviewed (changes requested):
N'oublie pas de squash les commits et puis on est good
